### PR TITLE
refact scan string

### PR DESCRIPTION
**Pull request by tiancaiamao**
refact a function `scanString`, make the code clean, and more efficient

### CODE REVIEW

File: parser/lexer.go [hunk 1]
```diff
@@ -285,70 +285,114 @@ func scanQuotedIdent(s *Scanner) (tok int, pos Pos, lit string) {
 }
 
 func startString(s *Scanner) (tok int, pos Pos, lit string) {
-	return s.scanString()
+	var quote bool
+	if s.r.peek() == '\'' {
```
Comment (shenli) [inline parser/lexer.go]: How about " and `?

Comment (tiancaiamao) [inline parser/lexer.go]: this is not a valid token...

File: parser/lexer.go [hunk 2]
```diff
@@ -285,70 +285,107 @@ func scanQuotedIdent(s *Scanner) (tok int, pos Pos, lit string) {
...
+	for ch == '\'' || ch == '"' {
+		_, _, lit1 := s.scanString()
+		lit = lit + lit1
+		ch = s.skipWhitespace()
+	}
+	return
+}
+
+// maybeBuf is used to avoid buf if possible.
+type maybeBuf struct {
```
Comment (shenli) [inline parser/lexer.go]: lazyBuffer is better

**Comment by shenli** [comment]
LGTM
PTAL @coocood @zimulala

File: parser/lexer.go [hunk 3]
```diff
@@ -285,70 +285,107 @@ func scanQuotedIdent(s *Scanner) (tok int, pos Pos, lit string) {
...
 	return
 }
 
+// handleEscape handles the case in scanString when previous char is '\'.
+func handleEscape(s *Scanner) rune {
+	s.r.inc()
+	ch0 := s.r.peek()
+	/*
+		\" \' \\ \n \0 \b \Z \r \t ==> escape to one char
+		\% \_ ==> preverse both char
```
Comment (coocood) [inline parser/lexer.go]: preserve

File: parser/lexer.go [hunk 4]
```diff
@@ -285,70 +285,107 @@ func scanQuotedIdent(s *Scanner) (tok int, pos Pos, lit string) {
...
+		_, _, lit1 := s.scanString()
+		lit = lit + lit1
+		ch = s.skipWhitespace()
+	}
+	return
+}
+
+// lazyBuf is used to avoid buf if possible.
+type lazyBuf struct {
+	flag bool
```
Comment (coocood) [inline parser/lexer.go]: add comments about flag.

File: parser/lexer.go [hunk 5]
```diff
@@ -285,70 +285,107 @@ func scanQuotedIdent(s *Scanner) (tok int, pos Pos, lit string) {
...
+	ch := s.skipWhitespace()
+	for ch == '\'' || ch == '"' {
+		_, _, lit1 := s.scanString()
+		lit = lit + lit1
+		ch = s.skipWhitespace()
+	}
+	return
+}
+
+// lazyBuf is used to avoid buf if possible.
```
Comment (coocood) [inline parser/lexer.go]: Do you mean avoid buf allocation?

File: parser/lexer.go (hunk 4 above)
Comment (tiancaiamao) [inline parser/lexer.go]: avoid bytes.Buffer.String().
when bytes.Buffer.String() method is called, a new string object is allocated.
I use the bytes.Buffer to deal with escape, such as \r \n \t... if no escape exists, buff is not necessary, use str[i:j] is enough

File: parser/lexer.go [hunk 6]
```diff
@@ -285,70 +285,107 @@ func scanQuotedIdent(s *Scanner) (tok int, pos Pos, lit string) {
...
+
+func (mb *lazyBuf) data() string {
+	var lit string
+	if mb.flag {
+		lit = mb.b.String()
+	} else {
+		lit = mb.r.data(mb.p)
+		lit = lit[1 : len(lit)-1]
+	}
+	return lit
```
Comment (tiancaiamao) [inline parser/lexer.go]: it's here...  
if mb.flag is false.... scanString is a zero alloc function
@coocood

File: parser/lexer.go (hunk 4 above)
Comment (coocood) [inline parser/lexer.go]: add the comment in code.

File: parser/lexer.go [hunk 7]
```diff
@@ -285,70 +285,107 @@ func scanQuotedIdent(s *Scanner) (tok int, pos Pos, lit string) {
...
 }
 
 func (s *Scanner) scanString() (tok int, pos Pos, lit string) {
-	s.buf.Reset()
 	tok, pos = stringLit, s.r.pos()
+	mb := lazyBuf{false, &s.r, &s.buf, &pos}
 	ending := s.r.readByte()
```
Comment (coocood) [inline parser/lexer.go]: Shall we rename 'readByte' to 'readRune'?

Comment (tiancaiamao) [inline parser/lexer.go]: no need... the caller must guarantee the first letter is '\'' or '\"'...
readByte is better performance than readRune

File: parser/lexer.go [hunk 8]
```diff
@@ -285,70 +285,110 @@ func scanQuotedIdent(s *Scanner) (tok int, pos Pos, lit string) {
...
+	}
+	return
+}
+
+// lazyBuf is used to avoid allocation if possible.
+// it has a flag field indicates whether bytes.Buffer is necessary. if
+// the flag is false, we can avoid calling bytes.Buffer.String(), which
+// make a copy of data and cause allocation.
+type lazyBuf struct {
+	flag bool
```
Comment (coocood) [inline parser/lexer.go]: How about rename 'flag' to 'useBuf' and rename method 'useBuf' to 'writeBuf'?

**Comment by coocood** [comment]
LGTM

**Pull request merged**